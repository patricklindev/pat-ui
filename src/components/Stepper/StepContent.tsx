import React, { FC, ReactNode } from 'react';

export interface IStepContentProps {
  /** children must be React Element */
  children?: ReactNode;
  /** set customized css class */
  className?: string;
}

export const StepContent: FC<IStepContentProps> = (props) => {
  const { children } = props;

  return <>{children}</>;
};

export default StepContent;
