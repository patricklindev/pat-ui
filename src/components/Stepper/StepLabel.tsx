import React, { FC, ReactNode, useEffect } from 'react';

export interface IStepLableProps {
  /** children must be React Element */
  children?: ReactNode;
  /** set customized css class */
  className?: string;
  /** Render given optional child */
  optional?: ReactNode;
  /** Render given step icon */
  StepIconComponent?: ReactNode;
  /**  */
  index: number;
  /**  */
  active?: boolean;
  /**  */
  error?: boolean;
  /** Set the completed status of a step */
  completed?: boolean;
}

const errorLabel = (
  <svg
    width="22"
    height="19"
    viewBox="0 0 22 19"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 19H22L11 0L0 19ZM12 16H10V14H12V16ZM12 12H10V8H12V12Z"
      fill="#F44336"
    />
  </svg>
);

const completedLabel = (
  <svg
    width="25"
    height="24"
    viewBox="0 0 25 24"
    fill="white"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 0a12 12 0 1 0 0 24 12 12 0 0 0 0-24zm-2 17l-5-5 1.4-1.4 3.6 3.6 7.6-7.6L19 8l-9 9z"
      fill="#3F51B5"
    ></path>
  </svg>
);

const numLabel = (active: boolean = false, index: number) => {
  return (
    <svg
      width="25"
      height="24"
      viewBox="0 0 25 24"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      {active ? (
        <rect x="0.5" width="24" height="24" rx="12" fill="#3F51B5" />
      ) : (
        <rect
          x="0.5"
          width="24"
          height="24"
          rx="12"
          fill="black"
          fill-opacity="0.38"
        />
      )}
      {index >= 10 ? (
        <text x="3.5" y="17.5" fill="white">
          {index + 1}
        </text>
      ) : (
        <text x="7.5" y="17.5" fill="white">
          {index + 1}
        </text>
      )}
    </svg>
  );
};

const StepIcon = (
  index: number,
  error = false,
  active = false,
  completed = false
) => {
  if (error) return errorLabel;
  else if (completed) return completedLabel;
  else return numLabel(active, index);
};

export const StepLabel: FC<IStepLableProps> = (props) => {
  const {
    children,
    className,
    optional,
    StepIconComponent,
    index,
    active,
    error,
    completed,
  } = props;

  const StepIconProps = {
    index: index,
    active: active,
    error: error,
    completed: completed,
  };

  return (
    <>
      {StepIcon(index, error, active, completed)}
      {children}
    </>
  );
};

StepLabel.defaultProps = {
  completed: false,
  error: false,
};

export default StepLabel;
