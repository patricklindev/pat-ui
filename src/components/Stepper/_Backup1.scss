// // @mixin font-package($font-size, ){}
// //   }

// .all-container-sm {
//     width:10%;
//     height:fit-content;
// }

// .all-container-md {
//     width:50%;
//     height:fit-content;
// }

// .all-container-lg {
//     width:100%;
//     height:fit-content;
// }


// .center-main-body {
//     width:fit-content;
//     min-width:300px;
//     border-style:groove;
//     border-radius:12.5px;
//     padding:7.5%;
// }

// .flex-row-container {
//     width:100%;
//     display: flex;
//     margin-top:1%;
// }

// //COLOR OF STATUS

// .gray {
//     background-color: $gray-600
// }

// .green {
//     background-color: green;
// }

// .red {
//     background-color: red;
// }

// .orange {
//     background-color: orange
// }


// //DESCRIPTION FOR HTML ARRAY

// .progress-container {
//     margin-left:50px;
//     width:100%;
//   }

// .progress-description{
//     margin-left:5%;  
// }



// //STEPPERTYPE==CIRCLE
// //FOR WHEN STEPPERTYPE IS "CIRCLE"


//   .circle {
//     height: 25px;
//     width: 25px;
//     border-radius: 50%;
//     display: inline-block;
//     text-align:center;
//     z-index: 1;
//     color:$gray-100;
//     margin-left:10%;
//   }


// //STEPPERTYPE==SQUARE
// //FOR WHEN STEPPERTYPE IS "SQUARE"



// .square {
//     height: 25px;
//     width: 25px;
//     display: inline-block;
//     text-align:center;
//     border-radius: 20%;
//     color:$gray-100;
//     margin-left:10%;
// }


// // BUTTON

// .Stepper-Button-Left {
//     width:25%;
//     margin-left:10%;
// }

// .spacer {
//     width:100%;
// }

// .vertical-spacer {
//     width:30%;
// }

// .button-vertical-margin {
//     margin-left:1%;
//     background-color:white;
// }

// .Stepper-Button-Right {
//     width:25%;
//     margin-right:10%;
// }

// .Stepper-Button-Skip {
//     width:25%;
//     margin-right:1%;
// }

// .Stepper-Button-Reset {
//     width:25%;
//     margin:auto;
// }

// // LINEBAR

// .bar {
//     width:100%;
//     min-width:20%;
//     height:5px;
//     border-radius: 10px;
//     display: inline-block;
// }


// // FONT Style

// .font-variant-main{
//     font-size: $font-size-sm;
//     list-style: none;
//     text-align: center;
//     border-radius: 25px;
// }

// .font-variant-vertical{
//     font-size: $font-size-sm;
//     font-style:italic;
//     list-style: none;
//     text-align: justify;
// }

// .font-variant-row{
//     font-size: $font-size-sm;
//     font-style:italic;
//     list-style: none;
//     text-align: center;
// }

// .font-variant-error{
//     font-size: $font-size-sm;
//     list-style: none;
//     color:red;
//     text-indent: 1%;
//     text-align: center;
// }

// // .icon-area {
    
// // }

// .description-area {
//     width:275px;
//     transition:height 2s;
//     height:75px;
//     background-color: white;
//     margin-bottom:1%;
//     margin-left:1%;
//     z-index: 0;
// }


// .bar-area {
//     width:50%;
//     min-width:20%;
//     height:5px;
//     margin-left:5%;
// }

// .display-in-vertical {
//     margin-top:10%;
//     height:50px;
// }
// .display-in-row {
//     width:100%;
//     height:50px;
// }

// .stepper-display {
//    width:80%;
//    margin:auto;
//    text-align:center
// }

// // PASSED IN PROP EXAMPLES

// .component-display{
//     background-color:gray;
//     height:150px;
// }

// .component-display-inner {
//     width:80%;
// }
