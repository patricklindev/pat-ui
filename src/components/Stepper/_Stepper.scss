.stepper {
  list-style-type: none;
  padding-left: 0;
  &.horizontal {
    display: flex;
    align-items: center;
    // justify-content: space-between;
    .stepper__divider {
      flex-grow: 1;
      border-bottom: 1px solid $secondary;
      margin: 3px;
    }
    .stepper__divider:last-child {
      display: none;
    }
  }
  &.vertical {
    display: flex;
    flex-direction: column;
    // gap: 1rem;
    .stepper__divider {
      // position: relative;
      padding-left: 12px;
      // margin-top: 3px;
      .stepper__divider-child {
        height: 20px;
        border-left: 1px solid $secondary;
      }
    }
    .stepper__divider:last-child {
      display: none;
    }
  }
  &.nonLinear {
    .stepper__step:hover {
      cursor: pointer;
    }
  }
  .stepper__step {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    .stepper__icon {
      margin-right: 5px;
    }
    .stepper__label {
      color: $secondary;
      font-weight: 500;
      &.stepper__label-active,
      &.stepper__label-completed {
        color: $black;
      }

      &.stepper__label-error {
        color: $danger;
      }
      .stepper__label-optional {
        display: block;
        font-size: 0.5rem;
      }
    }
    .lineBreaker {
      flex-basis: 100%;
      height: 0;
      margin-top: 3px;
    }
    .stepper__content {
      // background-color: yellow;
      margin-left: 12px;
      border-left: 1px solid $secondary;
      padding-left: 20px;
      // order: 1;
    }
  }
}
