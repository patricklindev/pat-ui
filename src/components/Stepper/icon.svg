<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="0.5" width="24" height="24" rx="12" fill="#3F51B5"/>

</svg>

<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="0.5" width="24" height="24" rx="12" fill="black" fill-opacity="0.38"/>
</svg>

<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="0.5" width="24" height="24" rx="12" fill="#3F51B5"/>
<path d="M12 0a12 12 0 1 0 0 24 12 12 0 0 0 0-24zm-2 17l-5-5 1.4-1.4 3.6 3.6 7.6-7.6L19 8l-9 9z"></path>

</svg>