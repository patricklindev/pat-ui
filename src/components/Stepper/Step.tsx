import React, {
  FC,
  ReactNode,
  Children,
  ReactElement,
  cloneElement,
} from 'react';
import { classNames } from '../../utils/classNames';
import { IStepLableProps } from './StepLabel';

export interface IStepProps {
  /** children must be React Element */
  children?: ReactElement | ReactElement[];
  /** set customized css class */
  className?: string;
  index?: number;
  activeStep?: number;
  /**  */
  onClick?: () => void;
}

export const Step: FC<IStepProps> = (props) => {
  const { children, className, index, activeStep, onClick } = props;
  let styleClasses = classNames('stepper__step', className ? className : '');
  return (
    <>
      <li className={styleClasses} onClick={onClick}>
        {children
          ? Children.map(children, (child: ReactElement) =>
              cloneElement(child, {
                index: index,
                active: activeStep === index,
                // completed: activeStep > index,
              })
            )
          : children}
      </li>
      <div className="stepper__divider">
        <div className="stepper__divider-child"></div>
      </div>
    </>
  );
};

export default Step;
