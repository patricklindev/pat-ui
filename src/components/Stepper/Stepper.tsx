import React, { Children, cloneElement, FC, ReactElement } from 'react';
import { IStepProps } from './Step';

// https://mui.com/material-ui/react-stepper/
// Find the UI design reference in Figma Material UI Kit 1.0.2 (https://www.figma.com/file/e9mrJNqitnThCajlQvuyWv/Material-UI-Kit-1.0.2-(Copy)?node-id=787%3A0)
// Developers can feed an array of string description
// Developers can choose whether to skip a step
// Developers can click on the buttons to mark completion of steps
// Developers can dictate orientation of stepper (row vs column)
// Developers can feed an error.

export interface IStepperProps {
  /** children must be React Element */
  children?: ReactElement<IStepProps> | ReactElement<IStepProps>[];
  /** set customized css class */
  className?: string;
  /** Set current step, starts from 0 */
  activeStep?: number;
  /** Set orientation */
  orientation?: string;
}

/**
 * Steppers display progress through a sequence of logical and numbered steps.
 *
 * ```js
 * import { Stepper } from 'pat-ui'
 * ```
 */
export const Stepper: FC<IStepperProps> = (props) => {
  const { children, className, activeStep, orientation } = props;
  return (
    <ul>
      {children
        ? Children.map(children, (child: ReactElement) =>
            cloneElement(child, { activeStep })
          )
        : children}
    </ul>
  );
};

Stepper.defaultProps = {
  activeStep: 0,
  orientation: 'horizontal',
};

export default Stepper;
