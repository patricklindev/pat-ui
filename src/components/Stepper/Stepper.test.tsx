import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Stepper from './Stepper';
import Step from './Step';
import StepLabel from './StepLabel';

const steps = [
  'Select campaign settings',
  'Create an ad group',
  'Create an ad',
];

describe('Stepper', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Stepper>
        {steps.map((label, index) => {
          return (
            <Step key={label} index={index}>
              <StepLabel>{label}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render three Steps given three strings', () => {
    const { container } = render(
      <Stepper>
        {steps.map((label, index) => {
          return (
            <Step key={label} index={index}>
              <StepLabel index={0}>{label}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );

    expect(container.firstChild).toHaveClass('stepper');
    expect(screen.queryAllByRole('listitem').length).toBe(3);
  });

  it('should render given label in each step', () => {
    const { container } = render(
      <Stepper>
        {steps.map((label, index) => {
          return (
            <Step key={label} index={index}>
              <StepLabel index={0}>{label}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );
    expect(screen.queryByText('Select campaign settings')).toBeInTheDocument;
    expect(screen.queryByText('Create an ad group')).toBeInTheDocument;
    expect(screen.queryByText('Create an ad')).toBeInTheDocument;
  });

  it('should render horizontal Stepper in default', () => {
    const { container } = render(
      <Stepper>
        {steps.map((label, index) => {
          return (
            <Step key={label} index={index}>
              <StepLabel index={0}>{label}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );
    expect(container.querySelector('.vertical')).toBeNull();
  });

  it('should render vertical Stepper when given props', () => {
    const { container } = render(
      <Stepper orientation="vertical">
        {steps.map((label, index) => {
          return (
            <Step key={label} index={index}>
              <StepLabel index={0}>{label}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );
    expect(container.querySelector('.vertical')).toBeTruthy();
  });

  it('should skip a step when clicking skip button', () => {});

  it('should mark completion of a step when clicking complete button', () => {});

  it('should render an error icon when given props', () => {});
});
