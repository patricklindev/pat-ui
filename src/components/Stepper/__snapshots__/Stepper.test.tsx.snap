// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should match snapshot 1`] = `
<DocumentFragment>
  <ul
    class="stepper horizontal  "
  >
    <li
      class="stepper__step "
    >
      <div
        class="stepper__icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 25 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="#3F51B5"
            height="24"
            rx="12"
            width="24"
            x="0.5"
          />
          <text
            fill="white"
            text-anchor="middle"
            x="50%"
            y="70%"
          >
            1
          </text>
        </svg>
      </div>
      <div
        class="stepper__label stepper__label-active   "
      >
        Select campaign settings
      </div>
    </li>
    <div
      class="stepper__divider"
    >
      <div
        class="stepper__divider-child"
      />
    </div>
    <li
      class="stepper__step "
    >
      <div
        class="stepper__icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 25 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="black"
            fill-opacity="0.38"
            height="24"
            rx="12"
            width="24"
            x="0.5"
          />
          <text
            fill="white"
            text-anchor="middle"
            x="50%"
            y="70%"
          >
            2
          </text>
        </svg>
      </div>
      <div
        class="stepper__label    "
      >
        Create an ad group
      </div>
    </li>
    <div
      class="stepper__divider"
    >
      <div
        class="stepper__divider-child"
      />
    </div>
    <li
      class="stepper__step "
    >
      <div
        class="stepper__icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 25 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="black"
            fill-opacity="0.38"
            height="24"
            rx="12"
            width="24"
            x="0.5"
          />
          <text
            fill="white"
            text-anchor="middle"
            x="50%"
            y="70%"
          >
            3
          </text>
        </svg>
      </div>
      <div
        class="stepper__label    "
      >
        Create an ad
      </div>
    </li>
    <div
      class="stepper__divider"
    >
      <div
        class="stepper__divider-child"
      />
    </div>
  </ul>
</DocumentFragment>
`;
